From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Thibaut Sautereau <thibaut.sautereau@ssi.gouv.fr>
Date: Fri, 20 Sep 2019 14:02:42 +0200
Subject: [PATCH 060/103] slub: Extend init_on_free to slab caches with
 constructors

This is the remaining non-upstream part of SLAB_SANITIZE, which was a
partial port, from Daniel Micay, of the feature from PaX without the
default fast mode based on passing SLAB_NO_SANITIZE in
performance-critical cases that are not particularly security sensitive.

Signed-off-by: Thibaut Sautereau <thibaut.sautereau@ssi.gouv.fr>
[levente@leventepolyak.net: Adapt to kasan init_on_free with HW_TAGS changes]
Signed-off-by: Levente Polyak <levente@leventepolyak.net>
---
 mm/slab.h | 12 +++++++++---
 mm/slub.c | 18 ++++++++++++++++++
 2 files changed, 27 insertions(+), 3 deletions(-)

diff --git a/mm/slab.h b/mm/slab.h
index bf2f87acf5e3a..714d8333a5a99 100644
--- a/mm/slab.h
+++ b/mm/slab.h
@@ -687,13 +687,19 @@ static inline bool slab_want_init_on_alloc(gfp_t flags, struct kmem_cache *c)
 }
 
 static inline bool slab_want_init_on_free(struct kmem_cache *c)
 {
 	if (static_branch_maybe(CONFIG_INIT_ON_FREE_DEFAULT_ON,
-				&init_on_free))
-		return !(c->ctor ||
-			 (c->flags & (SLAB_TYPESAFE_BY_RCU | SLAB_POISON)));
+				&init_on_free)) {
+#ifndef CONFIG_SLUB
+		if (c->ctor)
+			return false;
+#endif
+		if (c->flags & (SLAB_TYPESAFE_BY_RCU | SLAB_POISON))
+			return false;
+		return true;
+	}
 	return false;
 }
 
 #if defined(CONFIG_DEBUG_FS) && defined(CONFIG_SLUB_DEBUG)
 void debugfs_slab_release(struct kmem_cache *);
diff --git a/mm/slub.c b/mm/slub.c
index 661fb994092d3..727c600c2a77e 100644
--- a/mm/slub.c
+++ b/mm/slub.c
@@ -2682,10 +2682,12 @@ bool slab_free_hook(struct kmem_cache *s, void *x, bool init,
 		 * Restore orig_size, otherwise kmalloc redzone overwritten
 		 * would be reported
 		 */
 		set_orig_size(s, x, orig_size);
 
+		if (s->ctor)
+			s->ctor(x);
 	}
 	/* KASAN might put x into memory quarantine, delaying its reuse. */
 	return !kasan_slab_free(s, x, init, still_accessible, false);
 }
 
@@ -2725,10 +2727,26 @@ bool slab_free_freelist_hook(struct kmem_cache *s, void **head, void **tail,
 			/*
 			 * Adjust the reconstructed freelist depth
 			 * accordingly if object's reuse is delayed.
 			 */
 			--(*cnt);
+
+			/* Objects that are put into quarantine by KASAN will
+			 * still undergo free_consistency_checks(), which
+			 * checks whether the freelist pointer is valid if it
+			 * is located after the object (see check_object()).
+			 * Since this is the case for slab caches with
+			 * constructors, we need to fix the freelist pointer
+			 * after init_on_free has overwritten it.
+			 *
+			 * Note that doing this for all caches (not just ctor
+			 * ones) would cause a GPF due to KASAN poisoning and
+			 * the way set_freepointer() eventually dereferences
+			 * the freepointer.
+			 */
+			if (slab_want_init_on_free(s) && s->ctor)
+				set_freepointer(s, object, NULL);
 		}
 	} while (object != old_tail);
 
 	return *head != NULL;
 }
-- 
2.55.0

